src/linear_algebra/invariant_basis_number.lean

```diff
@@ -0,0 +1,228 @@
+a finite type. There should be lemmas applying the invariant basis number property to each
+situation.
+
+The finite version of the invariant basis number property implies the inifite analogue, i.e., that
+`(ι →₀ R) ≃ₗ[R] (ι' →₀ R)` implies that `cardinal.mk ι = cardinal.mk ι'`. This fact (and its
+variants) should be formalized.
+
+## References
+
+* https://en.wikipedia.org/wiki/Invariant_basis_number
+
+## Tags
+
+free module, rank, invariant basis number, IBN
+
+-/
+
+noncomputable theory
+
+open_locale classical big_operators
+
+universes u v w
+
+section
+variables (R : Type u) [ring R]
+
+/-- We say that `R` has the invariant basis number property if `(fin n → R) ≃ₗ[R] (fin m → R)`
+    implies `n = m`. This gives rise to a well-defined notion of rank of a finitely generated free
+    module. -/
+class invariant_basis_number : Prop :=
+(eq_of_fin_equiv : ∀ {n m : ℕ}, ((fin n → R) ≃ₗ[R] (fin m → R)) → n = m)
+
+end
+
+section
+variables (R : Type u) [ring R] [invariant_basis_number R]
+
+lemma eq_of_fin_equiv {n m : ℕ} : ((fin n → R) ≃ₗ[R] (fin m → R)) → n = m :=
+invariant_basis_number.eq_of_fin_equiv
+
+lemma nontrivial_of_invariant_basis_number : nontrivial R :=
+begin
+  by_contra h,
+  refine zero_ne_one (eq_of_fin_equiv R _),
+  haveI := not_nontrivial_iff_subsingleton.1 h,
+  haveI : subsingleton (fin 1 → R) := ⟨λ a b, funext $ λ x, subsingleton.elim _ _⟩,
+  refine { .. }; { intros, exact 0 } <|> tidy
+end
+
+end
+
+section
+open finite_dimensional
+
+/-- A field has invariant basis number. This will be superseded below by the fact that any nonzero
+    commutative ring has invariant basis number. -/
+lemma invariant_basis_number_field {K : Type u} [field K] : invariant_basis_number K :=
+⟨λ n m e,
+  calc n = fintype.card (fin n) : eq.symm $ fintype.card_fin n
+     ... = findim K (fin n → K) : eq.symm $ findim_eq_card_basis (pi.is_basis_fun K (fin n))
+     ... = findim K (fin m → K) : linear_equiv.findim_eq e
+     ... = fintype.card (fin m) : findim_eq_card_basis (pi.is_basis_fun K (fin m))
+     ... = m                    : fintype.card_fin m⟩
+
+end
+
+/-!
+  We want to show that nontrivial commutative rings have invariant basis number. The idea is to
+  take a maximal ideal `I` of `R` and use an isomorphism `R^n ≃ R^m` of `R` modules to produce an
+  isomorphism `(R/I)^n ≃ (R/I)^m` of `R/I`-modules, which will imply `n = m` since `R/I` is a field
+  and we know that fields have invariant basis number.
+
+  We construct the isomorphism in two steps:
+  1. We construct the ring `R^n/I^n`, show that it is an `R/I`-module and show that there is an
+     isomorphism of `R/I`-modules `R^n/I^n ≃ (R/I)^n`.
+  2. We construct an isomorphism of `R/I`-modules `R^n/I^n ≃ R^m/I^m` using the isomorphism
+     `R^n ≃ R^m`.
+-/
+
+section
+variables {R : Type u} [comm_ring R] (I : ideal R) (ι : Type v)
+
+/-- `I^n` as an ideal of `R^n`. -/
+def ideal.pi : ideal (ι → R) :=
```

Comment:
I think that everything from here till, say, line 183 could move to a more general file on modules.

Reply:
I agree. Do you have a suggestion which file that could be?

Reply:
I would say maybe just somewhere after the definition of ideal. I don't think it depends on anything advanced.

Reply:
I moved the lemmas to what is hopefully a good place in `ideals.lean`.